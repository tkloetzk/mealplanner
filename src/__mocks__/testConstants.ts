// src/test/testConstants.ts
import { CategoryType, ServingSizeUnit } from "@/types/food";
import { Kid } from "@/types/user";
import { BREAKFAST, LUNCH, DINNER, DAYS_OF_WEEK } from "@/constants";
import { RANCH_OPTION } from "@/constants/meal-goals";
// Constant for selected day
export const SELECTED_DAY = DAYS_OF_WEEK[6];

// Mock Kids
export const MOCK_KIDS: Kid[] = [
  { id: "kid1", name: "Child One" },
  { id: "kid2", name: "Child Two" },
];

// Category Types
export const PROTEINS = "proteins";
export const FRUITS = "fruits";
export const VEGETABLES = "vegetables";
export const GRAINS = "grains";
export const OTHER = "other";

// Mock Foods
export const MOCK_FOODS = {
  proteins: [
    {
      id: "1",
      name: "Chicken Breast",
      category: PROTEINS,
      calories: 165,
      protein: 31,
      carbs: 0.1,
      fat: 3.6,
      meal: [BREAKFAST, LUNCH, DINNER],
      servingSize: "1",
      servingSizeUnit: "piece" as ServingSizeUnit,
      hiddenFromKids: false,
    },
    {
      id: "4",
      name: "Not a breakfast food",
      category: "proteins" as CategoryType,
      calories: 100,
      protein: 30,
      carbs: 0.1,
      fat: 34.6,
      meal: [LUNCH, DINNER],
      servingSize: "1",
      servingSizeUnit: "piece" as ServingSizeUnit,
    },
    // Add more protein foods
  ],
  fruits: [
    {
      id: "2",
      name: "Apple",
      category: "fruits" as const,
      calories: 95,
      protein: 0.5,
      carbs: 25,
      fat: 0.3,
      meal: [BREAKFAST, LUNCH, DINNER],
      servingSize: "1",
      servingSizeUnit: "piece" as ServingSizeUnit, // Explicitly type as ServingSizeUnit
    },
    // Add more fruit foods
  ],
  vegetables: [
    {
      id: "3",
      name: "Broccoli",
      category: "vegetables" as CategoryType,
      calories: 55,
      protein: 4,
      carbs: 11,
      fat: 0.6,
      meal: [LUNCH, DINNER],
      servingSize: "1",
      servingSizeUnit: "cup" as ServingSizeUnit,
    },
    // Add more vegetable foods
  ],
  other: [
    {
      id: "5",
      name: "Mixed Snack Plate",
      calories: 200,
      protein: 5,
      carbs: 25,
      fat: 10,
      servingSize: "1",
      servingSizeUnit: "plate" as ServingSizeUnit,
      category: "other" as CategoryType,
      meal: [BREAKFAST, LUNCH, DINNER, "snack"],
      hiddenFromChild: true,
    },
    {
      id: "6",
      name: "Visible Snack",
      calories: 150,
      protein: 3,
      carbs: 20,
      fat: 5,
      servingSize: "1",
      servingSizeUnit: "serving" as ServingSizeUnit,
      category: "other" as CategoryType,
      meal: [BREAKFAST, LUNCH],
      hiddenFromChild: false,
    },
  ],
  condiments: [RANCH_OPTION],
};
