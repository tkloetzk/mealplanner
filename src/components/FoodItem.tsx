import React, { useMemo } from "react";
import { Sliders, Edit } from "lucide-react";
import { NutriScore } from "./NutriScore";
import { CategoryType, Food, SelectedFood } from "@/types/food";
import { getFoodImageSource } from "@/utils/imageUtils";
import Image from "next/image";

interface FoodItemProps {
  food: Food;
  category: CategoryType;
  isSelected: boolean;
  index: number;
  selectedFoodInCategory: SelectedFood | null;
  onSelect: () => void;
  onServingClick: (e: React.MouseEvent<HTMLDivElement>) => void;
  onEditFood?: () => void;
}

const FoodItem: React.FC<FoodItemProps> = ({
  food,
  category,
  index,
  isSelected,
  selectedFoodInCategory,
  onSelect,
  onServingClick,
  onEditFood,
}) => {
  const imageSource = useMemo(() => getFoodImageSource(food), [food]);

  return (
    <div
      data-testid={`${category}-${index}`}
      className={`relative p-4 rounded-lg transition-all duration-200 cursor-pointer
        ${
          isSelected
            ? "bg-blue-100 ring-2 ring-blue-500"
            : "hover:bg-gray-50 bg-white"
        }`}
      onClick={onSelect}
    >
      <div className="flex justify-between items-start gap-4">
        <div className="flex items-start gap-3">
          {imageSource && (
            <div className="w-16 h-16 relative rounded-lg overflow-hidden flex-shrink-0">
              <Image
                src={imageSource}
                alt={food.name}
                fill
                style={{ objectFit: "cover" }}
              />
            </div>
          )}
          <div>
            <div className="flex items-center gap-2">
              <h4 className="font-medium">{food.name}</h4>
              {food.score && <NutriScore score={food.score} />}
            </div>
            <div className="text-sm text-gray-600 mt-1">
              {food.servingSize} {food.servingSizeUnit}
            </div>
          </div>
        </div>

        <div className="text-right">
          <div className="font-medium">{food.calories} cal</div>
          <div className="text-sm text-gray-600 mt-1">
            P: {food.protein}g | C: {food.carbs}g | F: {food.fat}g
          </div>
        </div>
      </div>

      {isSelected && selectedFoodInCategory && (
        <div className="flex items-center justify-between mt-3 pt-3 border-t">
          <div className="text-sm font-medium text-blue-600">
            {selectedFoodInCategory.servings} serving(s) â€¢{" "}
            {Math.round(selectedFoodInCategory.adjustedCalories)} cal total
          </div>
          <div className="flex items-center gap-2">
            {onEditFood && (
              <div
                onClick={(e) => {
                  e.stopPropagation();
                  onEditFood();
                }}
                className="p-2 rounded-full hover:bg-blue-200 transition-colors"
                title="Edit Food"
              >
                <Edit className="h-4 w-4 text-blue-600" />
              </div>
            )}
            <div
              onClick={(e) => {
                e.stopPropagation();
                onServingClick(e);
              }}
              className="p-2 rounded-full hover:bg-blue-200 transition-colors"
              title="Adjust Servings"
            >
              <Sliders className="h-4 w-4 text-blue-600" />
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default React.memo(FoodItem);
