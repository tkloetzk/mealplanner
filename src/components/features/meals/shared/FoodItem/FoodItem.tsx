import React, { useMemo } from "react";
import { Sliders, Edit } from "lucide-react";
import { NutriScore } from "@/components/features/nutrition/NutritionSummary/components/NutriScore";
import { CategoryType, Food, SelectedFood } from "@/types/food";
import { getFoodImageSource } from "@/utils/imageUtils";
import Image from "next/image";
import { EyeOff, Eye } from "lucide-react";

interface FoodItemProps {
  food: Food;
  category: CategoryType;
  isSelected: boolean;
  index: number;
  selectedFoodInCategory: SelectedFood | null;
  onSelect: () => void;
  onServingClick: (e: React.MouseEvent<HTMLDivElement>) => void;
  onEditFood?: () => void;
  isHidden: boolean;
  onToggleVisibility: () => void;
  showVisibilityControls?: boolean;
}

export const FoodItem = ({
  food,
  category,
  index,
  isSelected,
  selectedFoodInCategory,
  onSelect,
  onServingClick,
  onEditFood,
  isHidden,
  onToggleVisibility,
  showVisibilityControls = false,
}: FoodItemProps) => {
  const imageSource = useMemo(() => getFoodImageSource(food), [food]);

  return (
    <div
      data-testid={`${category}-${index}`}
      className={`relative p-4 rounded-lg transition-all duration-200 cursor-pointer
      ${
        isSelected
          ? "bg-blue-100 ring-2 ring-blue-500"
          : "hover:bg-gray-50 bg-white"
      }
      ${isHidden ? "opacity-50" : ""}`}
      onClick={onSelect}
    >
      <div className="flex justify-between items-start gap-4">
        <div className="flex items-start gap-3">
          {imageSource && (
            <div className="w-16 h-16 relative rounded-lg overflow-hidden flex-shrink-0">
              <Image
                src={imageSource}
                alt={food.name}
                fill
                style={{ objectFit: "cover" }}
              />
            </div>
          )}
          <div>
            <div className="flex items-center gap-2">
              <h4 className="font-medium">{food.name}</h4>
              {food.score && <NutriScore score={food.score} />}
              {isHidden && (
                <div
                  className="text-xs bg-yellow-100 text-yellow-700 px-2 py-0.5 rounded"
                  title="This food is hidden from children's view"
                >
                  Hidden
                </div>
              )}
            </div>
            <div className="text-sm text-gray-600 mt-1">
              {food.servingSize} {food.servingSizeUnit}
            </div>
          </div>
        </div>
        {showVisibilityControls && (
          <div
            className="absolute top-2 right-2 p-1 rounded-full hover:bg-gray-100"
            onClick={(e) => {
              e.stopPropagation();
              onToggleVisibility();
            }}
            title={isHidden ? "Show to child" : "Hide from child"}
          >
            {isHidden ? (
              <EyeOff className="h-4 w-4 text-gray-500" />
            ) : (
              <Eye className="h-4 w-4 text-gray-500" />
            )}
          </div>
        )}
        <div className="text-right">
          <div className="font-medium">{food.calories} cal</div>
          <div className="text-sm text-gray-600 mt-1">
            P: {food.protein}g | C: {food.carbs}g | F: {food.fat}g
          </div>
        </div>
      </div>

      {isSelected && selectedFoodInCategory && (
        <div className="flex items-center justify-between mt-3 pt-3 border-t">
          <div className="text-sm font-medium text-blue-600">
            {selectedFoodInCategory.servings} serving(s) â€¢{" "}
            {Math.round(selectedFoodInCategory.adjustedCalories)} cal total
          </div>
          <div className="flex items-center gap-2">
            {onEditFood && (
              <div
                onClick={(e) => {
                  e.stopPropagation();
                  onEditFood();
                }}
                className="p-2 rounded-full hover:bg-blue-200 transition-colors"
                title="Edit Food"
              >
                <Edit className="h-4 w-4 text-blue-600" />
              </div>
            )}
            <div
              onClick={(e) => {
                e.stopPropagation();
                onServingClick(e);
              }}
              className="p-2 rounded-full hover:bg-blue-200 transition-colors"
              title="Adjust Servings"
            >
              <Sliders className="h-4 w-4 text-blue-600" />
            </div>
          </div>
        </div>
      )}
    </div>
  );
};
