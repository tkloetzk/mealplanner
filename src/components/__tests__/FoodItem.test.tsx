// src/components/__tests__/FoodItem.test.tsx
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import FoodItem from "../FoodItem";
import { MOCK_FOODS, PROTEINS } from "@/constants/tests/testConstants";

describe("FoodItem Component", () => {
  const mockOnSelect = jest.fn();
  const mockOnServingClick = jest.fn();
  const mockOnEditFood = jest.fn();

  const defaultProps = {
    id: "1",
    food: MOCK_FOODS.proteins[0],
    category: PROTEINS,
    isSelected: false,
    index: 0,
    selectedFoodInCategory: null,
    onSelect: mockOnSelect,
    onServingClick: mockOnServingClick,
    onEditFood: mockOnEditFood,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("renders food item correctly", () => {
    render(<FoodItem {...defaultProps} />);

    // Check core rendering elements
    expect(screen.getByText(MOCK_FOODS.proteins[0].name)).toBeInTheDocument();
    expect(
      screen.getByText(`${MOCK_FOODS.proteins[0].calories} cal`)
    ).toBeInTheDocument();
  });

  it("handles selection correctly", () => {
    render(<FoodItem {...defaultProps} isSelected={true} />);

    // Verify selected state styling or elements
    const foodItem = screen.getByTestId(`${defaultProps.category}-0`);

    expect(foodItem).toHaveClass("bg-blue-100");
  });

  it("triggers onSelect when clicked", () => {
    render(<FoodItem {...defaultProps} />);

    const foodItem = screen
      .getByText(MOCK_FOODS.proteins[0].name)
      .closest("div");
    fireEvent.click(foodItem!);

    expect(mockOnSelect).toHaveBeenCalledTimes(1);
  });

  it("shows serving details when selected", () => {
    const selectedFood = {
      ...MOCK_FOODS.proteins[0],
      servings: 2,
      adjustedCalories: MOCK_FOODS.proteins[0].calories * 2,
      adjustedProtein: MOCK_FOODS.proteins[0].protein * 2,
      adjustedCarbs: MOCK_FOODS.proteins[0].carbs * 2,
      adjustedFat: MOCK_FOODS.proteins[0].fat * 2,
    };

    render(
      <FoodItem
        {...defaultProps}
        isSelected={true}
        selectedFoodInCategory={selectedFood}
      />
    );

    expect(screen.getByText(/2 serving\(s\)/)).toBeInTheDocument();
  });

  it("matches snapshot of unselected food", () => {
    const { asFragment } = render(<FoodItem {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("matches snapshot of selected food", () => {
    const { asFragment } = render(
      <FoodItem {...defaultProps} isSelected={true} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
