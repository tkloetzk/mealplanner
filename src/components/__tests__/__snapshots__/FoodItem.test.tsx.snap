// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoodItem Component matches snapshot of selected food 1`] = `
<DocumentFragment>
  <div
    class="relative p-4 rounded-lg transition-all duration-200 cursor-pointer
        bg-blue-100 ring-2 ring-blue-500"
    data-testid="proteins-0"
  >
    <div
      class="flex justify-between items-start gap-4"
    >
      <div
        class="flex items-start gap-3"
      >
        <div>
          <div
            class="flex items-center gap-2"
          >
            <h4
              class="font-medium"
            >
              Chicken Breast
            </h4>
          </div>
          <div
            class="text-sm text-gray-600 mt-1"
          >
            1 piece
          </div>
        </div>
      </div>
      <div
        class="text-right"
      >
        <div
          class="font-medium"
        >
          165 cal
        </div>
        <div
          class="text-sm text-gray-600 mt-1"
        >
          P: 31g | C: 0g | F: 3.6g
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FoodItem Component matches snapshot of unselected food 1`] = `
<DocumentFragment>
  <div
    class="relative p-4 rounded-lg transition-all duration-200 cursor-pointer
        hover:bg-gray-50 bg-white"
    data-testid="proteins-0"
  >
    <div
      class="flex justify-between items-start gap-4"
    >
      <div
        class="flex items-start gap-3"
      >
        <div>
          <div
            class="flex items-center gap-2"
          >
            <h4
              class="font-medium"
            >
              Chicken Breast
            </h4>
          </div>
          <div
            class="text-sm text-gray-600 mt-1"
          >
            1 piece
          </div>
        </div>
      </div>
      <div
        class="text-right"
      >
        <div
          class="font-medium"
        >
          165 cal
        </div>
        <div
          class="text-sm text-gray-600 mt-1"
        >
          P: 31g | C: 0g | F: 3.6g
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
